Handle block comment before Traits in derive attribute (issue 4984)

ytmimi opened the pull request:
Resolves #4984

When a block comment comes before the traits of a derive attribute
the string "#[derive(" is included as part of the the pre snippet, and as
a result gets duplicated.

For example:

    #[derive(/*Debug, */Clone)] -> #[derive(#[derive(/*Debug, */ Clone)]

Now the string "#[derive(" is trimmed from the start of the pre snippet
before looking for the pre comment.

calebcartwright commented:
Are we sure the issue is specific to derive attributes? Seems to me that the actual issue/root cause is somewhere upstream in the call stack, and we should try to fix the underlying issue instead of working around the side effects that issue causes.

e.g. what happens with:
```rust
#[foo(/*Debug, */Clone)]
struct Foo;
```

ytmimi commented:
@calebcartwright On ``rustfmt-nightly v1.4.37`` the example you gave doesn't get reformatted.

Looking at the ``rewrite`` method it's probably because we branch if we're dealing with a derive (line 431)

https://github.com/rust-lang/rustfmt/blob/cb144c35e7d7151bb4b6ccd9b30a4141959166d1/src/attr.rs#L379-L494

I also totally agree that the problem should be fixed higher up in the call stack. I had trouble figuring out where a good place to fix the issue would be though. From my understanding this is what's happening:

``[<ast::Attribute] as Rewrite>::rewrite()`` calls [attr::format_derive](https://github.com/rust-lang/rustfmt/blob/cb144c35e7d7151bb4b6ccd9b30a4141959166d1/src/attr.rs#L93). ``attr::format_derive`` immediately maps each ``ast::Attribute`` into a [lists::ListItems](https://github.com/rust-lang/rustfmt/blob/cb144c35e7d7151bb4b6ccd9b30a4141959166d1/src/lists.rs#L556) object and then iterates and flattens the iterator of ``lists::ListItems`` into an iterator of [lists::ListItem](https://github.com/rust-lang/rustfmt/blob/cb144c35e7d7151bb4b6ccd9b30a4141959166d1/src/lists.rs#L122).

Taking a look at the [Iterator implementation for lists::ListItems](https://github.com/rust-lang/rustfmt/blob/cb144c35e7d7151bb4b6ccd9b30a4141959166d1/src/lists.rs#L721), which is where we construct each ``lists::ListItem``, we use the span to to find the Pre-Snippet, and then we use the Pre-Snippet to find the ``pre_comment``.

https://github.com/rust-lang/rustfmt/blob/cb144c35e7d7151bb4b6ccd9b30a4141959166d1/src/lists.rs#L730-L738

Given the following:

```rust
#[derive(/*Debug, */Clone)]
struct Foo;
```
The pre-snippet is ``"#[derive(/*Debug, */"`` and because the pre-snippet ends with a ``"*/"`` [lists::extract_pre_comment](https://github.com/rust-lang/rustfmt/blob/cb144c35e7d7151bb4b6ccd9b30a4141959166d1/src/lists.rs#L572) hits the following code block and just returns the trimmed pre-snippet as the found comment.

https://github.com/rust-lang/rustfmt/blob/cb144c35e7d7151bb4b6ccd9b30a4141959166d1/src/lists.rs#L591-L596

Let me know if all that makes sense.

Again, I wasn't really sure where the fix should go, but I thought it made the most sense to implement the fix in ``lists::extract_pre_comment``.

calebcartwright commented:
> Looking at the rewrite method it's probably because we branch if we're dealing with a derive (line 431)

Ah okay, thanks for confirming!

> The pre_snippet is "#[derive(/*Debug, */" and because the pre_snippet ends with a "*/" lists::extract_pre_comment hits the following code block and just returns the trimmed pre_snippet as the found comment.

Yeah it sounds like the starting point of the span is at the start of the attribute instead of the paren `(`. I feel like that's what we need to really fix, as carrying around extra baggage on formatting strings can be a source of other subtle bugs around width limit/wrapping checks.

ytmimi commented:
I removed the old fix where I just stripped ``"#[derive("`` from the start of the snippet and now I've updated the span to start after the opening ``"("``

calebcartwright commented on src/attr.rs:
```diff
@@ -116,7 +116,9 @@ fn format_derive(
-                attr.span.lo(),
+                // We update derive attribute spans to start after the opening '('
+                // This helps us focus parsing to just what's inside #[derive(...)]
+                attr.span.lo() + BytePos("#[derive(".len() as u32),
```
Excellent, this is precisely what I had in mind! This type of offset derivation is so common that we actually have a lot of utilities that make this easy to do, and account for various edge cases and complexities (whitespace, comments, etc.) so we'll want to apply the same idea here, but utilize the helper instead. We can also just look for the opening paren in this case, so that we don't have to worry about anything weird but technically valid like `#[ derive    (` type scenaro

e.g. (disclaimer, typed inline so may need to be tweaked and/or have the SpanUtils imported from source_map)
```suggestion
                context.snippet_provider.span_after(attr.span, "("),
```

calebcartwright commented:
Realize this and PR 5000 are still pending, but if you have any interest and bandwidth, I think #5009 and #5011 might be decent issues to look at as they're likely caused (at least in part) by the same sort of span/byte position derivation issues as this one

ytmimi commented:
Thanks for pointing out the helper method and for the heads up on bringing SpanUtils into scope! I went ahead and made the change.

I think your hunch about the other issues having a similar cause is probably right. I should have some time to take a crack at it later this week!

calebcartwright commented:
LGTM thanks!
